library/alloc/src/boxed.rs

```diff
@@ -62,6 +62,11 @@
 //! T` obtained from [`Box::<T>::into_raw`] may be deallocated using the
 //! [`Global`] allocator with [`Layout::for_value(&*value)`].
 //!
+//! For zero-sized values, the `Box` pointer still has to be [valid] for reads and writes and
+//! sufficiently aligned. In particular, casting any aligned non-zero integer literal to a raw
+//! pointer produces a valid pointer, but a pointer pointing into previously allocated memory that
+//! since got freed is not valid.
```

Comment:
Are there any limitations here -- i.e., it's only a problem if the integer value is somehow derived from a freed value via some trackable provenance -- or is it UB even if you just pull an integer out of thin air that *happens* to be the same as a freed value?

If the latter, can we give some concrete advice about what value people *should* use for zero-sized values? I at least am feeling a bit confused about what kind of value *is* valid to use, given that it can't be null but must be aligned.

Reply:
These are good questions. Unfortunately, LLVM is not specified precisely enough to properly answer them. We are building on quicksand here and can just take a guess about how it will behave. ;)

My personal interpretation is that it is okay to do this if the pointer has no provenance. That should always be the case for pointers created from an integer literal, or any other computations where there are no pointers involved.

Reply:
How would you feel about adding *some* text to that effect, even if it is quite tentative? What does the *compiler* do if we create a `Box::new(T)` where `struct T;` has some kind of high alignment?

Reply:
So you are saying I should use the term "provenance"? I had avoided that, because I don't think we are defining it anywhere outside the UCG glossary.

> What does the compiler do if we create a Box::new(T) where struct T; has some kind of high alignment?

This is implemented in the stdlib actually, not the compiler. It uses the alignment as ptr value:

https://github.com/rust-lang/rust/blob/c4f836ad1aceb83507810d9499f56988fd24578d/library/alloc/src/alloc.rs#L163

Reply:
No, I am not suggesting you should use the word provenance necessarily. What I was suggesting is saying something like "The recommended way to build a `Box` to a ZST is to use an integer equal to the alignment as the pointer value."

Reply:
And/or we could mention [`ptr::NonNull::dangling()`](https://doc.rust-lang.org/core/ptr/struct.NonNull.html#method.dangling) as a canonically-valid-and-always-up-to-date implementation 🙂

Reply:
That would be even better, good point.

Reply:
I added a comment along these lines.